File: java/Arrays/Day12/prodSubarray.java
package Day12;

import java.util.Scanner;

public class prodSubarray {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the size of Array: ");
        int n = sc.nextInt();

        System.out.print("Enter " + n + " " + "numbers: ");

        int arr[] = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }


        long maxProd = arr[0];
        long minProd = arr[0];
        long result = arr[0];

        for (int i = 1; i < n; i++) {
            long curr = arr[i];

            if (curr < 0) {
                long temp = maxProd;
                maxProd = minProd;
                minProd = temp;
            }

            maxProd = Math.max(curr, maxProd * curr);
            minProd = Math.min(curr, minProd * curr);

            result = Math.max(result, maxProd);
        }

        System.out.print("Maximum product subarray: " + result);


        sc.close();

    }
}

File: python/Arrays/day12/prodSubarray.py
n = int(input("Enter the size of array: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))

if not arr:
	print('Maximum product subarray: 0')
	exit()
	        
	        
maxProd = arr[0]
minProd = arr[0]
result = arr[0]
		
for i in range(1, len(arr)):
	curr = arr[i]
		    
	if curr < 0:
		maxProd, minProd = minProd, maxProd
		       
		    
	maxProd = max(curr, maxProd * curr)
	minProd = min(curr, minProd * curr)
		    
	result = max(result, maxProd)
		  
print("Maximum product subarray:", result)